Added ability to filter latest posts by author

pstonier opened the pull request:
## Description
Added dropdown to select author to filter the latest posts in the Latest Posts block.

## How has this been tested?
Locally.

## Screenshots <!-- if applicable -->
![Screen Shot 2019-06-13 at 4 50 27 PM](https://user-images.githubusercontent.com/2358202/59466484-72922980-8dfb-11e9-9f89-5f802d479fa4.png)
![Screen Shot 2019-06-13 at 4 50 39 PM](https://user-images.githubusercontent.com/2358202/59466485-72922980-8dfb-11e9-89e7-5f8150a91990.png)

## Types of changes
- Added AuthorSelect Component to QueryControls based on the CategorySelect Component
- Added `users` attribute to LatestPosts
- LatestPosts now fetches a user list to populate the dropdown like categories 

## Checklist:
- [ X ] My code is tested.
- [ X ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ X ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ X ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ n/a ] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

Commit:
Added ability to filter latest posts by author

talldan commented:
Thanks for working on this @pstonier!

The PR was discussed in a core-editor triage (https://wordpress.slack.com/archives/C02QB2JS7/p1562825386362400 - link requires slack sign up)

We decided to switch the label to Needs Design Feedback. Firstly we felt technical feedback can be provided through the normal code review process. Secondly, it'd be good to understand the motivation behind this PR (as there's no linked issue) and to make sure the UI is implemented in the right way in terms of design and behaviour of the dropdown.

pstonier commented:
All sounds good. From the motivation factor: we had been working on a page that was a bio about some of our consultants and needed a way to list the articles they'd authored. I had expected it to be an available filter option and when I found that it wasn't, I found my way around to adding it.

[Example Page](https://salienthealthcare.com/about/ryan-mackman/)

Commit:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

draganescu reviewed (commented):
This PR effectively copies the same logic as the category filter and from my tests it appears that it works correctly. I do not approve only because of the "Needs design feedback" label, but my opinion is that we go ahead with it.

paaljoachim commented:
This feature seems straight forward to me. 

Can some design folks add a comment please? As it would be nice to get it code reviewed and merged. @karmatosed @mapk @kjellr @jasmussen @melchoyce

jasmussen commented:
About to do some travelling so can't review in depth. But based on the screenshots alone, this seems good. It's an advanced feature, so it's in the sidebar, and it's a dropdown like the others. 👍 👍

paaljoachim commented:
This PR looks ready for a final code review and merge. Thanks!
@talldan

aprea commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -34,13 +34,17 @@ import { withSelect } from '@wordpress/data';
 	per_page: -1,
 };
+const USERS_LIST_QUERY = {
+	per_page: -1,
```
Wonder if this it going to scale okay? Just trying to think what'll happen on a site with hundreds of thousands or millions of users.

aprea replied:
Oh, never mind me just read the REST pagination docs.
> Large queries can hurt site performance, so per_page is capped at 100 records. If you wish to retrieve more than 100 records, for example to build a client-side list of all available categories, you may make multiple API requests and combine the results within your application.
https://developer.wordpress.org/rest-api/using-the-rest-api/pagination/

pstonier committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

pstonier committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

pstonier committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

pstonier committed:
Updates to catch up to changes in Gutenberg

draganescu commented:
In the mean time I've added #20415 implementing the same thing, with the hope of getting a faster course :) @pstonier would you look and either move this or that one towards a merge?

pstonier committed:
Fixes error of: Notice: Undefined index: items in /wp-includes/rest-api.php on line "1245"

pstonier committed:
Removed unused variables and fixed linting issues

pstonier committed:
Removed Whitespace from index.php

pstonier committed:
adding new line to index.php

mapk commented:
@draganescu what's the verdict here? Which PR are we going with?

draganescu commented:
Will review this one today

draganescu commented on packages/block-library/src/latest-posts/block.json:
```diff
@@ -12,6 +12,12 @@
+		"users": {
+			"type": "array"
+		},
+		"items": {
```
What do we need this items for? It seems that you needed it for the array type, but it should be nested into users. Now that I're recommented switching to number it is not needed at all.

pstonier replied:
Items was an attempt to resolve the issue of the error: "Notice: Undefined index: items in /app/public/wp-includes/rest-api.php on line 1245, but as you mentioned. It's not necessary and I've removed it.

I've also switched the type to number on users and is working as expected.

draganescu commented on packages/block-library/src/latest-posts/block.json:
```diff
@@ -12,6 +12,12 @@
 			"type": "array"
 		},
+		"users": {
+			"type": "array"
```
This makes the PR not work in my tests. Given that we only filter by one author and all we need is the ID which is a Number, we should have this set to "number"

pstonier replied:
I agree. I've updated the type to number.

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -262,6 +280,13 @@ class LatestPostsEdit extends Component {
 						selectedCategories={ categories }
+						onAuthorChange={ ( value ) =>
+							setAttributes( {
+								users: '' !== value ? value : undefined,
```
We should make sure we save the ID as number, it is string by default, so:
`users: '' !== value ? Number( value ) : undefined`

pstonier replied:
I agree. I've updated the code.

draganescu reviewed (commented):
Howdy @pstonier this didn't work out of the box. I left some comments about what I had to change. Other than that, great work, we're almost done!

pstonier committed:
applied changes suggested by @draganescu to LatestPosts block

pstonier committed:
linting edit

pstonier committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -97,6 +114,7 @@ class LatestPostsEdit extends Component {
 			order,
 			orderBy,
 			categories,
+			users,
```
As above, let's update this to author.

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -262,6 +280,14 @@ class LatestPostsEdit extends Component {
 						selectedCategories={ categories }
+						onAuthorChange={ ( value ) =>
+							setAttributes( {
+								users:
```
This as well.

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -459,6 +485,7 @@ export default withSelect( ( select, props ) => {
 		order,
 		orderBy,
 		categories,
+		users,
```
Same, update to author.

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -470,6 +497,7 @@ export default withSelect( ( select, props ) => {
 	const latestPostsQuery = pickBy(
 		{
 			categories: catIds,
+			author: users,
```
we can change to shorthand here:
 ```
{
	categories: catIds,
	author: users,
...
```

draganescu commented on packages/block-library/src/latest-posts/index.php:
```diff
@@ -49,6 +49,9 @@ function render_block_core_latest_posts( $attributes ) {
 		$args['category__in'] = array_column( $attributes['categories'], 'id' );
 	}
+	if ( isset( $attributes['users'] ) ) {
+		$args['author'] = $attributes['users'];
```
`$args['author'] = $attributes['author'];`

draganescu reviewed (changes requested):
Hey @pstonier we're 99% there, only one request left, which I should have spotted the last time I reviewed, that is using author instead of users as an attribute (and hence updating all over accordingly). It makes more sense, both the singular and the name change.

pstonier committed:
Updated from users to author on LatestPosts block

pstonier commented on packages/block-library/src/latest-posts/block.json:
```diff
@@ -12,6 +12,9 @@
 		"categories": {
 			"type": "array"
 		},
+		"users": {
```
I'm good with this.

mapk reviewed (approved):
I just tested this today and it's working as expected. I like that I'm able to filter the posts by author now! Thanks. I'm approving from a design perspective.

One thing to note is now that I can filter by Author, I'd like a toggle to show the author which should probably exist here:

![Screen Shot 2020-05-05 at 8 05 15 AM](https://user-images.githubusercontent.com/617986/81081756-31898480-8ea7-11ea-8429-fce0150ca089.png)

Of course, don't let that hold up this PR.

pstonier committed:
Added display author option

pstonier commented:
> I just tested this today and it's working as expected. I like that I'm able to filter the posts by author now! Thanks. I'm approving from a design perspective.
> 
> One thing to note is now that I can filter by Author, I'd like a toggle to show the author which should probably exist here:
> 
> ![Screen Shot 2020-05-05 at 8 05 15 AM](https://user-images.githubusercontent.com/617986/81081756-31898480-8ea7-11ea-8429-fce0150ca089.png)
> 
> Of course, don't let that hold up this PR.

I like that idea and have updated the code to add that.

pstonier committed:
Updated edit.js on LatestPosts to match front end on author label

pstonier committed:
Attempt to fix tests on LatestPosts

pstonier committed:
Update e2e tests for LatestPostsa

pstonier committed:
Fix for PHP Unit tests

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -411,6 +446,12 @@ class LatestPostsEdit extends Component {
 								</a>
+								{ displayAuthor && (
+									<div className="wp-block-latest-posts__post-author">
+										{ __( 'Posted by' ) }{ ' ' }
```
👋 This should be incorporating the author's name like in #20595 

```JS
<span className="wp-block-latest-posts__post-author">
	{ sprintf(
		/* translators: byline. %s: current author.
		__( 'by %s' ),
		currentAuthor.name
	) }
</span>
```

pstonier replied:
I agree this is a better approach. I've updated my code accordingly.

draganescu commented:
@pstonier @mapk  #20595 adds the "Show post author" toggle in the inspector.

pstonier committed:
Updated rendering of author name on editor view

mapk commented:
When the code is dialed in again and you'd like another review, just ping me. @draganescu thanks for stepping in to help.

pstonier committed:
Changed 'posted by' to 'by' on the byline

pstonier committed:
Fixed linting issue in index.php for LatestPost

pstonier committed:
Fixed formatting issues in index.php for LatestPosts

pstonier committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

draganescu commented:
Howdy @pstonier ,

Trying to test this, setting show author option I get:

```JS
TypeError: post.author_info is undefined edit.js:453:15
```

and the block crashes. 

Also a rebase would be in order.

pstonier committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/latest-posts-author-filter

pstonier committed:
fix to undefined author_info

pstonier commented:
> Howdy @pstonier ,
> 
> Trying to test this, setting show author option I get:
> 
> ```js
> TypeError: post.author_info is undefined edit.js:453:15
> ```
> 
> and the block crashes.
> 
> Also a rebase would be in order.

Thanks. I have fixed this and rebased.

draganescu reviewed (approved):
I have tested this and works pretty well. Since we're only adding new data to the block we shouldn't need a deprecation function, testing a previously created block with this PR updated with no issues.

Great job @pstonier !

pstonier commented:
Thank you for all the help!
